File: Core CS/Algorithms/Implementation/Sales by Match/salesByMatch.java
import java.io.*;
import java.util.*;
import java.text.*;

public class Solution {

    static int sockMerchant(int n, int[] ar) {
         Hashtable<Integer, Integer> m = new Hashtable<Integer, Integer>();
         int r = 0;

         for(int a: ar){
             if(m.containsKey(a))
                m.put(a, m.get(a) + 1);
            else
                m.put(a, 1);
         }

         for(int a: m.values())
            r += a/2;

        return r;
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[] ar = new int[n];
        for(int ar_i = 0; ar_i < n; ar_i++){
            ar[ar_i] = in.nextInt();
        }
        int result = sockMerchant(n, ar);
        System.out.println(result);
    }
}

File: Core CS/Algorithms/Implementation/Sales by Match/salesByMatch.kt
import java.io.*
import java.text.*
import java.util.*
import kotlin.text.*

fun sockMerchant(n: Int, ar: Array<Int>): Int {
    return ar.groupingBy{it}.eachCount().values.fold(0){ acc, b -> acc + b/2}
}

fun main(args: Array<String>) {
    val scan = Scanner(System.`in`)

    val n = scan.nextLine().trim().toInt()

    val ar = scan.nextLine().split(" ").map{ it.trim().toInt() }.toTypedArray()

    val result = sockMerchant(n, ar)

    println(result)
}
